File: ws-09-overview-and-classification-basic-patterns-immutable-object/src/main/java/com/arohau/creational/singleton/Singleton.java
package com.arohau.creational.singleton;

public class Singleton {
    private Singleton() {}

    /*
    private static class SingletonHolder {
        public static final Singleton INSTANCE = new Singleton();
    }
     */

    private static class SingletonHolder {
        public static final Singleton INSTANCE;
        static {
            System.out.println("here");
            INSTANCE = new Singleton();
        }
    }

    public static Singleton getInstance() {
        return SingletonHolder.INSTANCE;
    }

    public Object doSomething() {
        return "work is done";
    }
}

File: typescript/ws-10-basic-patterns-singleton-n-factory-method-n-abstract-factory/singleton/ex1/src/Singleton.ts
export class Singleton {
    private static instance: Singleton;

    // The constructor should be private to prevent direct construction calls with the `new` operator.
    private constructor() {
        // Initialization code here
    }

    // This static method controls the access to the singleton instance.
    public static getInstance(): Singleton {
        // On the first run, create the instance and save it to the static property.
        if (!Singleton.instance) {
            Singleton.instance = new Singleton();
        }

        // Return the singleton instance.
        return Singleton.instance;
    }

    public someMethod() {
        // Example method that can be called on the Singleton.
        console.log('Executing some process...');
    }
}
